### PR TITLE
implemented GroupTimeSeriesSplit with tests and documentation

### LINKED ISSUE #537: GroupTimeSeriesSplit
**Please explain clearly what you'd like to see added.** 

- To do a KFold cross validation with a time series data set is a bit more complex than other use cases. Folds have to retain chronological order. On top of that it is possible that not every time unit (e.g.) years has the same amount of rows/observations. Therefore, in case of a data set which contains observations over 100 years, it is not feasible to simply split the data set into KFolds with each fold containing the same amount of years, because the first 20 years might contain only 5% of the observations while the last 20% could contain 60% of the observations. 

**add a screenshot if applicable (ML stuff is hard to explain with words, pictures say 1000 words)**
image
_If the amount of observations increase over the years, the data should e.g. be split in these folds to have en equal amount of rows/observations per fold._

**make sure that the feature you want is not already supported by sklearn**
- To equally split the data based on the amount of observations, while keeping the chronological order of the time unit is not a KFold method in scikit-learn. scikit-learn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html#sklearn.model_selection.TimeSeriesSplit) assumes equal group size.

### PR DESCRIPTION
**Pull request by sTomerG**
# Description

**Fixes #537** 

To do a KFold cross validation with a time series data set is a bit more complex than other use cases. Folds have to retain chronological order. On top of that it is possible that not every time unit (e.g.) years has the same amount of rows/observations. Therefore, in case of a data set which contains observations over 100 years, it is not feasible to simply split the data set into KFolds with each fold containing the same amount of years, because the first 20 years might contain only 5% of the observations while the last 20% could contain 60% of the observations.

image

_If the amount of observations increase over the years, the data should e.g. be split in these folds to have en equal amount of rows/observations per fold._

**Make sure that the feature you want is not already supported by sklearn**

To equally split the data based on the amount of observations, while keeping the chronological order of the time unit is not a KFold method in scikit-learn. scikit-learn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html#sklearn.model_selection.TimeSeriesSplit) assumes equal group size.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] My code follows the style guidelines (flake8)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (also to the readme.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added tests to check whether the new feature adheres to the sklearn convention
- [x] New and existing unit tests pass locally with my changes

If you feel your PR is ready for a review, ping @koaning or @mbrouns.

### CODE REVIEW

File: sklego/common.py [hunk 1]
```diff
@@ -166,3 +166,27 @@ def expanding_list(list_to_extent, return_type=list):
         return [listed]
 
     return [return_type(listed[: n + 1]) for n in range(len(listed))]
+
+
+def chunker(sequence, chunk_size, step_size):
```
Comment (koaning) [inline sklego/common.py:171]: Wouldn't `sliding_window` be a better name for this?

Comment (koaning) [reply sklego/common.py:171]: I also wouldn't mind a test for this, nothing fancy but can't hurt to have.

File: sklego/model_selection.py [hunk 2]
```diff
@@ -244,7 +267,382 @@ def split(self, X, y=None, groups=None):
...
+                .assign(
+                    diff_from_ideal_group_size=lambda df: df["obs_per_group"]
+                    - df["ideal_group_size"]
+                )
+            )
+        except AttributeError:
+            raise AttributeError(
+                ".summary() only works after having ran"
+                " .split(X, y, groups)."
+            )
```
Comment (koaning) [inline sklego/model_selection.py:354]: I wouldn't mind having such a utility function in the documentation, but does it have to be a part of the public API? Do other sklearn components implement this?

**Review by koaning — COMMENTED** [review]
Only had time for a very quick review, hope to have more time later!

File: sklego/model_selection.py (hunk 2 above)
Comment (sTomerG) [reply sklego/model_selection.py:354]: `TimeGapSplit` of scikit-lego also has a method called `.summary()` which returns a `pd.DataFrame`, which I used as inspiration for this method.  I do believe this method is very helpful and informative for users to help understand the final group division.

File: sklego/common.py (hunk 1 above)
Comment (sTomerG) [reply sklego/common.py:171]: Sure, I'll implement both suggestions!

Comment (sTomerG) [reply sklego/common.py]: My colleague @davidelagano whom I worked with on `GroupTimeSeriesSplit` has implemented the suggestions!

File: tests/test_common/test_basics.py [hunk 3]
```diff
@@ -1,4 +1,6 @@
-from sklego.common import as_list
+import hypothesis.strategies as st
+from hypothesis import given
```
Comment (koaning) [inline tests/test_common/test_basics.py:2]: Do we really need hypothesis here? It'd be my preference to omit it.

File: sklego/common.py [hunk 4]
```diff
@@ -166,3 +166,27 @@ def expanding_list(list_to_extent, return_type=list):
...
+    :param chunk_size: the size of each chunk
+    :type chunk_size: int
+    :param step_size: the amount of steps to the next chunk
+    :type step_size: int
+    :return: a generator object
+    :rtype: Generator
+
+    :Example:
+
+    >>> generator = chunker([1,2,4,5], 2, 1)
```
Comment (koaning) [inline sklego/common.py:185]: The docs still list the name `chunker`.

**Review by koaning — CHANGES_REQUESTED** [review]
I prefer not to introduce a library just for a unit test in this PR. 

Saw some minor issues, which I commented but this is starting to look good! 

@MBrouns any strong impressions on your end?

**Comment by MBrouns** [comment]
Huh I was convinced we were already using Hypothesis somewhere. I think for this simple use-case it's indeed not worth it to introduce

**Comment by sTomerG** [comment]
All feedback should be implemented now!

File: doc/crossvalidation.ipynb [hunk 5]
```diff
@@ -247,7 +247,7 @@
...
...bfmMiP2AM4F/3J3ra12N5POcyt9Dt0eE//+s7nz+Eji7sn0WMCwiRu5O2+2picI0IvYB7gDm7vQv4VRGPjr91/BK0fAZWv+HdhAwNCIuLChcdaFG8jkf+DAwgdZfHt/q5f77jBrJ5xeBfxsRzbROw/lDL/ffZ9RIPsnM72Tmh4H/DPx1b/ffV9RKPivOB27PzHdK6r901ZbPyqjPQuD6zFxdVD/VqkbyeQ8wNjOPoXXk7ocdXV/raiSfXwVOiIingROAF4Ee+51b9YVpRAyiNcm3ZuZPK7vXV6blbZ+et6GLZj4FPJeZLZn5FvBT4OMR8dE2D2tPp/Xmt/2XnjGVfR3FNqDN9d+onNt2akOn1/dHtZLPzFyfme9k5rvAjcBx3bsDtaWG8rkyM0/LzMm0/pJ+tnt3oLZUWT676+/op1PwazCf59MPp/FuV6X5XACsysx5u/JZ+4NayWdmbsrMNytvvwdM7uqz16Iayue6zDw7MycC/6Wyr8cWnBvYUw2VISIC+D6wIjOva3PobuALwDWVnz/roqnngSkRMQTYBpwCNGXmE7SOeG3vbwTw3+OP0zJPAy7vqNHKv9pOaLsvIrZGxBRah/D/HPjfXX3O/qKW8hkRB7aZnnEWsKyLmGtOjeVz/8zcEBF/Quvo2ne7iLnmVGM+O/ksh2XmqsrbPwVWdXZ+LaqlfFbaP4LWRT4e6+41taQa8xkR3wT2BfrdKq1dqaV87vT30HRan6/sV2osn6OA31cGXi4Hbuoi5l2TfWC1qt19AZ+gddh7KbCk8joDGEnrXOhVwEPAiMr5B9A6z3orsKWyPbxy7CpgJa0FxC3A3h30+UXgN5XXRW32/02lvXcrP6/s4PrGSh/PAn8LRGX/WZXr3gTWA4vKvr/mc4/yeQvwq8pnuRs4sOz7az73KJ9fpnUVvV/T+j+QKPv+ms9u5bPd84Bv0/rc8BJaF29oKPv+ms/dz2fl2JXANWXfV/PZvXzSOoKTtBYp2+O9uHLsI5XrXgM2AcvLvr/mc4/y+T9o/X37S1p/3x5R9v01n3uUz3Mr8f6a1hHwdvvf3df2P7okSZIkSSpF1T9jKkmSJEmqbhamkiRJkqRSWZhKkiRJkkplYSpJkiRJKpWFqSRJkiSpVBamkiRJkqRSWZhKkiRJkkr1/wEuNaFP79DgzgAAAABJRU5ErkJggg==\n",
```
Comment (MBrouns) [inline doc/crossvalidation.ipynb:250]: I guess Jupyter changed how images are stored (`\n` at the end). This line change seems unnecessary, but we'll have this with every new update to the notebook. @koaning are you okay with updating this ipynb to the new style in this PR?

Comment (koaning) [reply doc/crossvalidation.ipynb:250]: That's annoying.

But no biggy. Feel free to update.

**Comment by sTomerG** [comment]
I'm applying it in a project now, something seems off, I'll have to double check.

**Comment by koaning** [comment]
Thanks for letting us know. 

And sorry for the delay in the review! Life with a newborn (and it's diapers) sure is ... different 😓

**Comment by sTomerG** [comment]
> Thanks for letting us know.
> 
> And sorry for the delay in the review! Life with a newborn (and it's diapers) sure is ... different 😓

No worries I understand!

Found the cause of the unexpected behavior, have to make a minor change and will add some tests to check for it. I'll push the code today or tomorrow! In the mean time I'm also writing a Medium article to explain the GroupTimeSeriesSplit, which I guess I'll post when the merge is completed :)

**Comment by sTomerG** [comment]
I have [this article](https://medium.com/@tomergabay/a-highly-anticipated-time-series-cross-validator-is-finally-here-7dc99f672736) about this cross-validator ready for publishing, which I'll submit to Towards Data Science when the pull request is merged :)

File: sklego/model_selection.py [hunk 6]
```diff
@@ -244,7 +267,394 @@ def split(self, X, y=None, groups=None):
...
+            return True
+        except NotFittedError:
+            return False
+
+
+class GroupTimeSeriesSplit(_BaseKFold):
+    """
+    Sliding window time series split
+
+    - Create n_splits folds with an as equally possible size through a smart variant of a brute
```
Comment (koaning) [inline sklego/model_selection.py:291]: Nitpick: do we need the `-` here?

File: tests/test_model_selection/test_grouptimeseriessplit.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+@pytest.fixture
+def valid_cv(init_data, valid_n_splits):
+    X, y, groups = init_data
+    cv = GroupTimeSeriesSplit(valid_n_splits)
+    _ = cv.split(X, y, groups)
+    return cv
+
+
+@pytest.mark.parametrize("n_splits", [(1), ("a"), (2.2), ([5]), (-1)])
+def test_invalid_n_splits(init_data, n_splits):
```
Comment (koaning) [inline tests/test_model_selection/test_grouptimeseriessplit.py:43]: Is `init_data` used in the test? If not, maybe remove it?

**Review by koaning — COMMENTED** [review]
I mainly had some nitpicks. 

@MBrouns any objections from your end? If not I'm good to merge this one and release a new 0.6.x version.

**Comment by MBrouns** [comment]
No more feedback from my end! Looks good to merge after your nitpicks are resolved

**Comment by DavideLagano** [comment]
> No more feedback from my end! Looks good to merge after your nitpicks are resolved

Nitpicks are resolved.

**Comment by koaning** [comment]
Grand! I've just merged it and will make a release later today.

**Comment by sTomerG** [comment]
I believe the documentation is not yet updated? @koaning

**Comment by koaning** [comment]
My bad! 

Just pushed tho 

https://scikit-lego.netlify.app/crossvalidation.html#GroupTimeSeriesSplit

**Pull request merged**